### PR TITLE
Dense 2 softmax

### PR DESCRIPTION
**Pull request by rainarashika**

### CODE REVIEW

File: dqn_lstm.py [hunk 1]
```diff
@@ -90,9 +92,34 @@ def replay(self, batch_size):
...
 
-    def predict(self, *args, **kwargs):
-        # Directly call the predict method of the encapsulated Keras model
-        return self.model.predict(*args, **kwargs)
+    def predict(self, X):
+        logits = self.model(X, training=False).numpy()  # Model already outputs probabilities due to softmax
+
+        if self.calibration_method == 'none':
+            logits = self.model(X, training=False).numpy()
+            probabilities = tf.nn.softmax(logits, axis=-1).numpy()
```
Comment (desimmons) [inline dqn_lstm.py:100]: Is the dimension of this `probabilities` vector definitely 2?

Comment (rainarashika) [reply dqn_lstm.py:100]: > Is the dimension of this `probabilities` vector definitely 2?

 yes, initially, the probabilities vector is of dimension 2 (i.e., it has 2 columns, one for each class), but after I selected the first column and reshaped it, the output now is just a single-column vector.

Comment (desimmons) [reply dqn_lstm.py:100]: I thought we saw the dimension of the output tensor from the model as `(4,2,1)`?

File: dqn_lstm.py [hunk 2]
```diff
@@ -41,7 +41,9 @@ def build_model(self):
         model = Sequential()
         model.add(LSTM(self.lstm_units, input_shape=(self.input_shape[1], self.input_shape[2]),return_sequences= False))
         model.add(Dense(10, activation='PReLU'))
-        model.add(Dense(1,activation='sigmoid'))  # Use the first element of output_shape as the number of units
+        #model.add(Dense(1,activation='sigmoid'))  # Use the first element of output_shape as the number of units
+        #model.add(Dense(2,activation='softmax'))
```
[the remark this answers is not in the document]
Comment (rainarashika) [reply dqn_lstm.py:45]: And, after applying softmax here, I didn't apply softmax later again (for uncalibrated). I just extracted probabilities for class 0.

File: dqn_lstm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rainarashika) [reply dqn_lstm.py:100]: I had tried including Softmax in training, but the results were still not the same. After applying softmax here, I didn't apply softmax later again (for uncalibrated). I just extracted probabilities for class 0.
(see line 44 that I commented)
